Two build warnings in CLANG DataFormats/Math/test/CholeskyInvert_t.cpp

In the Clang IB we have two warnings due to 

`  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/cff53344cac83c1abd88bb561d95fc5c/opt/cmssw/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_CLANG_X_2021-01-25-2300/src/DataFormats/Math/test/CholeskyInvert_t.cpp:48:6: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]`

which says clang unable to vectorize a loop 
probably around the pragmas in the same file 
https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/test/CholeskyInvert_t.cpp#L97

Full log is in:
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc900/CMSSW_11_3_CLANG_X_2021-01-25-2300/DataFormats/Math

One hint is there is a option to be enabled in our llvm build to enable clang to vectorize loops (have not tried yet if the problem is in the llvm external).

--- a/DataFormats/Math/interface/choleskyInversion.h
+++ b/DataFormats/Math/interface/choleskyInversion.h
@@ -337,7 +337,8 @@ namespace math {
 
     // Eigen interface
     template <typename D1, typename D2>
-    inline constexpr void __attribute__((always_inline)) invert(Eigen::DenseBase<D1> const& src, Eigen::DenseBase<D2>& dst) {
+    inline constexpr void __attribute__((always_inline))
+    invert(Eigen::DenseBase<D1> const& src, Eigen::DenseBase<D2>& dst) {
       using M1 = Eigen::DenseBase<D1>;
       using M2 = Eigen::DenseBase<D2>;
       Inverter<M1, M2, M2::ColsAtCompileTime>::eval(src, dst);

--- a/DataFormats/Math/interface/choleskyInversion.h
+++ b/DataFormats/Math/interface/choleskyInversion.h
@@ -20,13 +20,13 @@ namespace math {
   namespace cholesky {
 
     template <typename M1, typename M2>
-    inline constexpr void invert11(M1 const& src, M2& dst) {
+    inline constexpr void __attribute__((always_inline)) invert11(M1 const& src, M2& dst) {
       using F = decltype(src(0, 0));
       dst(0, 0) = F(1.0) / src(0, 0);
     }
 
     template <typename M1, typename M2>
-    inline constexpr void invert22(M1 const& src, M2& dst) {
+    inline constexpr void __attribute__((always_inline)) invert22(M1 const& src, M2& dst) {
       using F = decltype(src(0, 0));
       auto luc0 = F(1.0) / src(0, 0);
       auto luc1 = src(1, 0) * src(1, 0) * luc0;
@@ -40,7 +40,7 @@ namespace math {
     }
 
     template <typename M1, typename M2>
-    inline constexpr void invert33(M1 const& src, M2& dst) {
+    inline constexpr void __attribute__((always_inline)) invert33(M1 const& src, M2& dst) {
       using F = decltype(src(0, 0));
       auto luc0 = F(1.0) / src(0, 0);
       auto luc1 = src(1, 0);
@@ -64,7 +64,7 @@ namespace math {
     }
 
     template <typename M1, typename M2>
-    inline constexpr void invert44(M1 const& src, M2& dst) {
+    inline constexpr void __attribute__((always_inline)) invert44(M1 const& src, M2& dst) {
       using F = decltype(src(0, 0));
       auto luc0 = F(1.0) / src(0, 0);
       auto luc1 = src(1, 0);
@@ -100,7 +100,7 @@ namespace math {
     }
 
     template <typename M1, typename M2>
-    inline constexpr void invert55(M1 const& src, M2& dst) {
+    inline constexpr void __attribute__((always_inline)) invert55(M1 const& src, M2& dst) {
       using F = decltype(src(0, 0));
       auto luc0 = F(1.0) / src(0, 0);
       auto luc1 = src(1, 0);
@@ -155,7 +155,7 @@ namespace math {
     }
 
     template <typename M1, typename M2>
-    inline __attribute__((always_inline)) constexpr void invert66(M1 const& src, M2& dst) {
+    inline constexpr void __attribute__((always_inline)) invert66(M1 const& src, M2& dst) {
       using F = decltype(src(0, 0));
       auto luc0 = F(1.0) / src(0, 0);
       auto luc1 = src(1, 0);
@@ -297,7 +297,7 @@ namespace math {
 
     template <typename M1, typename M2>
     struct Inverter<M1, M2, 2> {
-      static constexpr void eval(M1 const& src, M2& dst) {
+      static constexpr void __attribute__((always_inline)) eval(M1 const& src, M2& dst) {
         invert22(src, dst);
         symmetrize22(dst);
       }
@@ -305,7 +305,7 @@ namespace math {
 
     template <typename M1, typename M2>
     struct Inverter<M1, M2, 3> {
-      static constexpr void eval(M1 const& src, M2& dst) {
+      static constexpr void __attribute__((always_inline)) eval(M1 const& src, M2& dst) {
         invert33(src, dst);
         symmetrize33(dst);
       }
@@ -313,7 +313,7 @@ namespace math {
 
     template <typename M1, typename M2>
     struct Inverter<M1, M2, 4> {
-      static constexpr void eval(M1 const& src, M2& dst) {
+      static constexpr void __attribute__((always_inline)) eval(M1 const& src, M2& dst) {
         invert44(src, dst);
         symmetrize44(dst);
       }
@@ -321,7 +321,7 @@ namespace math {
 
     template <typename M1, typename M2>
     struct Inverter<M1, M2, 5> {
-      static constexpr void eval(M1 const& src, M2& dst) {
+      static constexpr void __attribute__((always_inline)) eval(M1 const& src, M2& dst) {
         invert55(src, dst);
         symmetrize55(dst);
       }
@@ -329,7 +329,7 @@ namespace math {
 
     template <typename M1, typename M2>
     struct Inverter<M1, M2, 6> {
-      static constexpr void eval(M1 const& src, M2& dst) {
+      static constexpr void __attribute__((always_inline)) eval(M1 const& src, M2& dst) {
         invert66(src, dst);
         symmetrize66(dst);
       }
@@ -337,7 +337,7 @@ namespace math {
 
     // Eigen interface
     template <typename D1, typename D2>
-    inline constexpr void invert(Eigen::DenseBase<D1> const& src, Eigen::DenseBase<D2>& dst) {
+    inline constexpr void __attribute__((always_inline)) invert(Eigen::DenseBase<D1> const& src, Eigen::DenseBase<D2>& dst) {
       using M1 = Eigen::DenseBase<D1>;
       using M2 = Eigen::DenseBase<D2>;
       Inverter<M1, M2, M2::ColsAtCompileTime>::eval(src, dst);

--- a/DataFormats/Math/test/CholeskyInvert_t.cpp
+++ b/DataFormats/Math/test/CholeskyInvert_t.cpp
@@ -15,7 +15,6 @@
 
 #include "DataFormats/Math/interface/choleskyInversion.h"
 
-
 constexpr int stride() { return 5 * 1024; }
 template <int DIM>
 using MXN = Eigen::Matrix<double, DIM, DIM>;
@@ -108,6 +107,9 @@ void go(bool soa) {
     else
 #ifdef USE_VECTORIZATION_PRAGMA
 #pragma GCC ivdep
+#ifdef __clang__
+#pragma clang loop vectorize(enable) interleave(enable)
+#endif
 #endif
       for (auto& m : mm) {
         math::cholesky::invert(m, m);

--- a/DataFormats/Math/test/CholeskyInvert_t.cpp
+++ b/DataFormats/Math/test/CholeskyInvert_t.cpp
@@ -15,6 +15,7 @@
 
 #include "DataFormats/Math/interface/choleskyInversion.h"
 
+
 constexpr int stride() { return 5 * 1024; }
 template <int DIM>
 using MXN = Eigen::Matrix<double, DIM, DIM>;
@@ -94,16 +95,20 @@ void go(bool soa) {
   for (int kk = 0; kk < NKK; ++kk) {
     delta -= (std::chrono::high_resolution_clock::now() - start);
     if (soa)
+#ifdef USE_VECTORIZATION_PRAGMA
 #pragma GCC ivdep
 #ifdef __clang__
 #pragma clang loop vectorize(enable) interleave(enable)
+#endif
 #endif
       for (unsigned int i = 0; i < SIZE; ++i) {
         MapMX<N> m(p + i);
         math::cholesky::invert(m, m);
       }
     else
+#ifdef USE_VECTORIZATION_PRAGMA
 #pragma GCC ivdep
+#endif
       for (auto& m : mm) {
         math::cholesky::invert(m, m);
       }

